### PR TITLE
Feature #2138 CI enhance MET Docker tag override

### LINKED ISSUE #2138: Documentation: Improve info for CI/testing in GHA to force using a dev tag of the MET software
Per discussion with @georgemccabe on Slack:

```
Is there a way to run the METplus wrappers automation but using a specific branch of the MET repo? Or does this type of testing better suited for doing locally and confirming it runs before the METplus wrappers automation uses the new beta of the MET?
```

```
short answer yes, but not as easily as we would like. It is possible to force the METplus automated tests to use a specific Docker tag for MET by setting the MET_FORCE_TAG variable on this line of the workflow file:
https://github.com/dtcenter/METplus/blob/108abbca5044cc07528c441bf20d61ae50742980/.github/workflows/testing.yml#L88

However, that only corresponds to existing MET Docker tags, not GitHub branches. It looks like Docker tags for PRs are created on DockerHub dtcenter/met-dev, but the logic uses the dtcenter/met if you set MET_FORCE_TAG because it was assumed that you would set this to use an older, stable release of MET in this case.

Line that sets MET DockerHub repo:
https://github.com/dtcenter/METplus/blob/108abbca5044cc07528c441bf20d61ae50742980/.github/jobs/docker_setup.sh#L49

It looks like there is a DockerHub image for [dtcenter/met:feature_2285_read_point_data-PR](https://hub.docker.com/layers/dtcenter/met-dev/feature_2285_read_point_data-PR/images/sha256-ac45cd0ce0842ff8bacd1b925dca42766d4f53f5c2e1de9591adbdc72d4a1c1f?context=repo), so you should be able to use that branch if you:
Change line 49 in .github/jobs/docker_setup.sh to MET_DOCKER_REPO=met-dev
Set MET_FORCE_TAG=feature_2285_read_point_data-PR on line 88 in .github/workflows/testing.yml

It would be nice to enhance the automated test logic so it is easier to handle this case without having to change multiple files.
```

### Time Estimate ###
1 hour

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

Tested setting SET_MET_IMAGE to `met:11.0.0` and `met-dev:feature_2597_disk_space` and confirmed that it properly built the METplus Docker image using the correct MET images

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Review documentation updates: https://metplus.readthedocs.io/en/feature_2138_force_met_tag/Contributors_Guide/continuous_integration.html#force-met-version-used-for-tests

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[No]**</br>

- [X] Please complete this pull request review by **7/25/2023**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Add any new Python packages to the [METplus Components Python Requirements](https://metplus.readthedocs.io/en/develop/Users_Guide/overview.html#metplus-components-python-requirements) table.
- [X] Review the source issue metadata (required labels, projects, and milestone).
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: docs/Contributors_Guide/continuous_integration.rst [hunk 1]
```diff
@@ -496,15 +496,33 @@ development testing. If testing is done on a stable release, then the
...
+
+Another situation that may require a different MET Docker image is if there
+are changes in a MET feature or bugfix branch that are needed to test changes
+in METplus.
+
+To allow the tests to run successfully in these cases, an option was added to
+force a specific MET Docker image to be used to build the METplus Docker image
+that is used for testing.
+
+In the **testing.yml** workflow file, there is a commented variable called
```
Comment (DanielAdriaansen) [inline docs/Contributors_Guide/continuous_integration.rst:511]:
```suggestion
In the **testing.yml (.github/workflows/testing.yml)** workflow file, there is a commented variable called
```

File: docs/Contributors_Guide/continuous_integration.rst [hunk 2]
```diff
@@ -496,15 +496,33 @@ development testing. If testing is done on a stable release, then the
...
+SET_MET_IMAGE that can be uncommented and set the MET Docker image to use.
+This variable is found in the **get_image** job under the **env** section
 for the step named "Get METplus Image."
```
Comment (DanielAdriaansen) [inline docs/Contributors_Guide/continuous_integration.rst:514]:
```suggestion
for the step named "Get METplus Image in the **testing.yml** file."
```

File: docs/Contributors_Guide/continuous_integration.rst [hunk 3]
```diff
@@ -496,15 +496,33 @@ development testing. If testing is done on a stable release, then the
...
 
+The format of the value is <REPO>:<TAG>
+where the DockerHub repo used is dtcenter/<REPO> and the tag used is <TAG>.
+
+Stable releases of MET are found in the dtcenter/met DockerHub repo and
```
Comment (DanielAdriaansen) [inline docs/Contributors_Guide/continuous_integration.rst]: Where is the "dtcenter/met DockerHub repo"?

File: docs/Contributors_Guide/continuous_integration.rst [hunk 4]
```diff
@@ -496,15 +496,33 @@ development testing. If testing is done on a stable release, then the
...
+are named using the X.Y.Z version of the release,
+so setting **SET_MET_IMAGE=met:11.1.0** will use dtcenter/met:11.1.0.
+
+Development versions of MET are found in the dtcenter/met-dev DockerHub
```
Comment (DanielAdriaansen) [inline docs/Contributors_Guide/continuous_integration.rst]: Similarly, should we include a URL to the "dtcenter/met-dev DockerHub repo" here?

**Review by DanielAdriaansen — COMMENTED** [review]
A few suggestions and questions. Otherwise looks good. Happy to iterate.

File: docs/Contributors_Guide/continuous_integration.rst (hunk 1 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst:511]: The "GitHub Actions Workflows" section mentions that all of the workflow files are found in .github/workflows. The testing.yml file is mentioned many times before this section, so I don't think it is necessary to list the path here. I could put a note under the "9.2 Testing Workflow" header to list the path to the file if that is desired?

File: docs/Contributors_Guide/continuous_integration.rst (hunk 2 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst:514]: The sentence before this mentions that the variable is in testing.yml. Mentioning it again here seems redundant.

File: docs/Contributors_Guide/continuous_integration.rst (hunk 3 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst]: On DockerHub here: https://hub.docker.com/repository/docker/dtcenter/met/general

File: docs/Contributors_Guide/continuous_integration.rst (hunk 4 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst]: https://hub.docker.com/repository/docker/dtcenter/met-dev/general
I will add links for this and the above DockerHub repo.

File: docs/Contributors_Guide/continuous_integration.rst (hunk 1 above)
Comment (DanielAdriaansen) [reply docs/Contributors_Guide/continuous_integration.rst:511]: I think the references you note are sufficient. Thanks for pointing them out.

File: docs/Contributors_Guide/continuous_integration.rst (hunk 3 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst]: Added link

File: docs/Contributors_Guide/continuous_integration.rst (hunk 4 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst]: Added link

File: docs/Contributors_Guide/continuous_integration.rst [hunk 5]
```diff
@@ -496,15 +498,35 @@ development testing. If testing is done on a stable release, then the
...
+In the **testing.yml** workflow file, there is a commented variable called
+SET_MET_IMAGE that can be uncommented and set the MET Docker image to use.
+This variable is found in the **get_image** job under the **env** section
 for the step named "Get METplus Image."
 
+The format of the value is <REPO>:<TAG>
+where the DockerHub repo used is dtcenter/<REPO> and the tag used is <TAG>.
+
+Stable releases of MET are found in the
+`dtcenter/met DockerHub repo <hub.docker.com/repository/docker/dtcenter/met/general>`_
```
Comment (DanielAdriaansen) [inline docs/Contributors_Guide/continuous_integration.rst]: IDK why, but this URL doesn't work. Here's where it points in the rendered docs:
https://metplus.readthedocs.io/en/feature_2138_force_met_tag/Contributors_Guide/hub.docker.com/repository/docker/dtcenter/met-dev/general

**Review by DanielAdriaansen — CHANGES_REQUESTED** [review]
Something weird with URL.

File: docs/Contributors_Guide/continuous_integration.rst (hunk 5 above)
Comment (georgemccabe) [reply docs/Contributors_Guide/continuous_integration.rst]: Ah, I need to include https or it assumes it is relative to the ReadTheDocs pages

**Review by DanielAdriaansen — APPROVED** [review]
Looks good, thanks for making these changes @georgemccabe!

**Pull request merged**